[Frontend][Backend] Add device-side tma descriptor update API

peterbell10 opened the pull request:
This adds several new operators to the nvidia gpu IR which expose
1. loading a descriptor into shared memory
2. updating different fields on the descriptor
3. copying the descriptor back to global memory
4. fencing operations on the descriptor

I then use these to expose 3 new functions in `tl.extra.cuda`.
1. `experimental_device_tensormap_create1d`
2. `experimental_device_tensormap_create2d`
3. `experimental_tensormap_fenceproxy_acquire`

which allows for new tensormap descriptor creation based on a template, and gives the fence operation needed to use the updated descriptor. We also have the tools needed to implement partial updates if they end up being needed for performance reasons.

peterbell10 commented on include/triton/Dialect/Triton/IR/TritonOps.td:
```diff
@@ -1133,7 +1134,7 @@ def TT_ExperimentalDescriptorLoadOp : TT_Op<"experimental_descriptor_load", [
 
 def TT_ExperimentalDescriptorStoreOp : TT_Op<"experimental_descriptor_store", [
-  MemoryEffects<[MemWrite<GlobalMemory>]>]> {
+  MemoryEffects<[MemRead<GlobalMemory>, MemWrite<GlobalMemory>]>]> {
```
Drive-by change. We read the TMA descriptor from global memory so this has a read effect.

peterbell10 commented on python/test/unit/language/test_core.py:
```diff
@@ -15,9 +15,22 @@
+    numpy_random,
+    to_triton,
+    torch_dtype_name,
+    to_numpy,
```
I refactored some of these helpers so they can be shared with the tma test file, there are no functional changes.

peterbell10 commented on third_party/nvidia/backend/driver.c:
```diff
@@ -306,6 +306,7 @@ static PyObject *fill1DTMADescriptor(PyObject *self, PyObject *args) {
     break;
   default:
     PyErr_SetString(PyExc_ValueError, "elementSize must be 1, 2, or 4");
+    return NULL;
```
Another drive-by. `PyErr_SetString` requires the caller to return `NULL` to propagate the exception into python.

Jokeren commented on include/triton/Dialect/Triton/IR/Dialect.h:
```diff
@@ -27,6 +27,10 @@ struct GlobalMemory : public SideEffects::Resource::Base<GlobalMemory> {
 };
 
+struct SharedMemory : public SideEffects::Resource::Base<GlobalMemory> {
+  StringRef getName() final { return "<GlobalMemory>"; }
```
It's weird to me if shared memory effects are inherited from global memory. 
1. Does shared memory mean GPU shared memory or something else?
2. If the former, exposing it in TritonOps seems like breaking our semantics because TritonOps are device agnostic.
3. If the later, can we change a name?
4. Also, why shared memory's `getName` method returns `<GlobalMemory>`?

peterbell10 replied:
Whoops, this was copy-pasted from `GlobalMemory` definition above and I forgot to completely fixup.

> Does shared memory mean GPU shared memory or something else?

Yes this is GPU shared memory.

> If the former, exposing it in TritonOps seems like breaking our semantics because TritonOps are device agnostic.

I could try to come up with some way to hide the fact these are using shared memory, but at the end of the day TMA descriptors are already a highly device specific, so I'm not sure there's any benefit there.

peterbell10 replied:
And I should add that the triton language level doesn't actually see the shared memory, it only exists in the frontend when building the IR.

Jokeren replied:
> I could try to come up with some way to hide the fact these are using shared memory, but at the end of the day TMA descriptors are already highly device specific, so I'm not sure there's any benefit there.

Got it

Jokeren commented on include/triton/Dialect/Triton/IR/TritonOps.td:
```diff
@@ -1156,4 +1157,204 @@ def TT_ExperimentalDescriptorStoreOp : TT_Op<"experimental_descriptor_store", [
     }];
 }
 
+def TT_ExperimentalTensormapAllocOp: TT_Op<
```
It seems to me that it not only has the `Alloc` side effect, but also has the `Write` side effect.

ThomasRaoux commented on python/triton/language/extra/cuda/_experimental_tma.py:
```diff
@@ -0,0 +1,116 @@
+    global_size = core._to_tensor(global_size, _builder)
+    element_ty = core._constexpr_to_value(element_ty)
+
+    local_desc = semantic.tensormap_allocate(template_desc, _builder)
+    semantic.tensormap_replace_global_address(local_desc, global_address, _builder)
+    _set_tma_dtype(local_desc, element_ty, _builder)
+    semantic.tensormap_replace_rank(local_desc, 0, _builder)
+    semantic.tensormap_replace_box_dim(local_desc, 0, load_size, _builder)
+    semantic.tensormap_replace_global_dim(local_desc, 0, global_size, _builder)
+    semantic.tensormap_replace_element_stride(local_desc, 0, 1, _builder)
+    semantic.tensormap_replace_interleave_layout(local_desc, 0, _builder)
+    semantic.tensormap_replace_swizzle_mode(local_desc, 0, _builder)
+    semantic.tensormap_replace_fill_mode(local_desc, 0, _builder)
+
+    semantic.tensormap_cp_fenceproxy(desc_out, local_desc, builder=_builder)
+    semantic.tensormap_deallocate(local_desc, _builder)
```
I think this kind of lowering would fit better in TTGIR, this would avoid exposing as much details to TTIR and would allow better re-use of the existing TTGIR operations instead of having to add so many new ops to TTIR.

We could just have a TTIR op that maps `experimental_device_tensormap_create` kind of signature then lower it in TTGIR to a shared memory alloc with some local_stores

ThomasRaoux commented on lib/Analysis/Alias.cpp:
```diff
@@ -34,8 +34,9 @@ LogicalResult SharedMemoryAliasAnalysis::visitOperation(
-  // Only LocalAllocOp creates a new buffer.
-  if (isa<triton::gpu::LocalAllocOp>(op)) {
+  // New buffer creation ops.
+  if (isa<triton::gpu::LocalAllocOp, triton::ExperimentalTensormapAllocOp>(
```
I find it a bit weird to have to create a new allocation op just for TMA. Can we lower to the existing `LocalAllocOp`?

ThomasRaoux commented on python/triton/language/extra/cuda/_experimental_tma.py:
```diff
@@ -0,0 +1,116 @@
+    global_size = core._to_tensor(global_size, _builder)
+    element_ty = core._constexpr_to_value(element_ty)
+
+    local_desc = semantic.tensormap_allocate(template_desc, _builder)
```
I know I had mentioned we should use templates before but since we are updating most of the fields anyway I wonder if we should just create the descriptor from scratch as it will most likely be more efficient. (See the discussion we had with Elliot on slack)

ThomasRaoux reviewed (commented):
Nice work! 
Added a few comments mostly on layering of ops as I think we could have less ops and layer it in a way that would match better the amount of HW details we usually expose at TTIR/TTGIR

ThomasRaoux commented on python/triton/language/extra/cuda/_experimental_tma.py:
```diff
@@ -0,0 +1,116 @@
+    global_stride = semantic.mul(element_size_t, global_size[-1], _builder)
+    # Undocumented, but global_stride seems to be divided by 16
+    global_stride = semantic.ashr(global_stride, core._to_tensor(4, _builder), _builder)
+    semantic.tensormap_replace_global_stride(local_desc, 0, global_stride, _builder)
+
+    contig_dim_size_in_bytes = element_size * load_size[-1]
+    if contig_dim_size_in_bytes > 128:
+        load_size[-1] = 128 // element_size
+
+    for irank in range(rank):
+        # 0th element in tensormap corresponds to this last dimension of the tensor
+        idim = rank - (irank + 1)
+        semantic.tensormap_replace_box_dim(local_desc, irank, load_size[idim], _builder)
+        semantic.tensormap_replace_global_dim(local_desc, irank, global_size[idim], _builder)
+        semantic.tensormap_replace_element_stride(local_desc, irank, 1, _builder)
+
+    swizzle = _determine_swizzle_mode_2d(contig_dim_size_in_bytes, load_size)
+    semantic.tensormap_replace_interleave_layout(local_desc, 0, _builder)
+    semantic.tensormap_replace_swizzle_mode(local_desc, swizzle, _builder)
+    semantic.tensormap_replace_fill_mode(local_desc, 0, _builder)
+
+    semantic.tensormap_cp_fenceproxy(desc_out, local_desc, builder=_builder)
+    semantic.tensormap_deallocate(local_desc, _builder)
```
Note that the descriptor layout is described [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#encoding-a-tensor-map-on-device):
```
struct tensormap_params {
  void* global_address;
  int rank;
  uint32_t box_dim[5];
  uint64_t global_dim[5];
  size_t global_stride[4];
  uint32_t element_stride[5];
};
```
It might be simpler and more efficient to directly writes this out?

peterbell10 commented:
@ThomasRaoux as discussed, I've:
1. Changed TTIR to only have a single `tt.experimental_tensormap_create` op that doesn't expose shared memory.
2. Moved fine grained tensormap ops to the nvidia IR
3. Updated LocalAllocOp to take an alignment attribute and removed TensormapAlloc

ThomasRaoux commented:
> @ThomasRaoux as discussed, I've:
> 
> 1. Changed TTIR to only have a single `tt.experimental_tensormap_create` op that doesn't expose shared memory.
> 2. Moved fine grained tensormap ops to the nvidia IR
> 3. Updated LocalAllocOp to take an alignment attribute and removed TensormapAlloc

Thanks!

> 2. Moved fine grained tensormap ops to the nvidia IR

do we even need those ops. What if we were lowering `tt.experimental_tensormap_create` directly to llvm?

peterbell10 commented:
@ThomasRaoux PTAL, removed nvidia ops and now lower directly to LLVM. Also removed the read of the template and now just zero-fill the smem descriptor.

ThomasRaoux reviewed (approved):
Looks great! 
One optional extra nit: maybe add a simple lit test for the lowering of the new op, it can be convenient to easily check the kind of code it lowers to, it's not super important though so feel free to skip

ThomasRaoux reviewed (approved):
Thanks!
